Support setting FDT entry in extlinux.conf

20k-ultra opened the pull request:
- [x] Ensure test coverage for previous functionality
- [x] Separate config backends into their own files
- [x] Implement sorting target config variables by their entry type
- [x] Implement a APPEND interface so APPEND entry type variables can be set (uses existing code)
- [x] Implement a FDT interface so FDT entry type variables can be set

Closes: #1207 
Signed-off-by: Miguel Casqueira <miguel@balena.io>

CameronDiver commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,86 @@
+	// Parses the values for this directive
+	public abstract parse(directives: Directive): ConfigOptions;
+	// Return the value to be set for this directive using the provided ConfigOptions
+	public abstract value(opts: ConfigOptions, existingValue: string): string;
```
At the risk of bike-shedding here, I'm not super fond of this method name. Something a little more applicable, such as `generate` or `compose` might be nice (although compose might be confused with docker-compose).

20k-ultra replied:
I totally agree. Compose is perfect but in the off chance it can be confused with docker-compose I'll opt for generate.

CameronDiver commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,86 @@
+	[directive: string]: string;
+}
+
+export abstract class ConfigurableDirective {
```
Given that we were talking about the ease of creating new directives, could you add a couple of comments here detailing that this is the base class of the directives, to make a new one, implement this etc etc.

Basically just an overview of what its for for any future contributors.

20k-ultra replied:
I've added documentation to the class abstraction and each directive class.

CameronDiver commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,86 @@
+}
+
+export interface Directive {
+	[directive: string]: string;
```
We could more strongly type this, right? 
```
export interface Directive {
    [key in 'APPEND' | 'FDT']: string
```

20k-ultra replied:
I have been able to achieve a stronger typing for Directive but it actually just complicated things more. I may need some help to figure out a way to do this.

20k-ultra replied:
We agreed to move on from this for now as there's no decent solution I liked.

CameronDiver commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,86 @@
+
+		for (const value of values) {
+			const parts = value.split('=');
+			if (parts.length !== 2) {
```
Did you look at whether the values could contain a `=` too? How is that handled. Given we're testing for example 2 entries, that could cause issues. I expect they're just not allowed, but it's worth checking.

20k-ultra replied:
To my knowledge a key within the APPEND value cannot have more then one =.

APPEND ro rootwait isolcpu=2,3 other=4 // OK
APPEND ro rootwait isolcpu=2,3other=4 // WRONG

Therefore your comment has made me aware of an issue where the parse wouldn't error on these invalid values. I have modified it to catch any config options that have more then a KEY and VALUE.

```js
const [KEY, VALUE, ...more] = appendValue.split('=');
if (more.length > 0) {
// APPEND value is not formatted correctly
}
```

CameronDiver commented on src/config/backends/extlinux.ts:
```diff
@@ -248,4 +232,31 @@ export class ExtlinuxConfigBackend extends DeviceConfigBackend {
+
+	private static findDefaultLabel(file: ExtlinuxFile): string {
+		const defaultLabel = _.find(file.globals, (_v, l) => {
+			if (l === 'DEFAULT') {
```
```suggestion
			return label === 'DEFAULT'
```

20k-ultra replied:
added.

CameronDiver commented on src/lib/errors.ts:
```diff
@@ -103,6 +103,12 @@ export class ContractViolationError extends TypedError {
 }
 
+export class ExtLinuxParseError extends TypedError {
+	public constructor(parseError: string) {
```
You can omit the constructor when it only calls `super`.

20k-ultra replied:
fixed!

Page- commented on src/config/backends/raspberry-pi.ts:
```diff
@@ -0,0 +1,157 @@
+
+	public processConfigVarValue(key: string, value: string): string | string[] {
+		if (_.includes(RPiConfigBackend.arrayConfigKeys, key)) {
+			if (!_.startsWith(value, '"')) {
```
If we know it's a string then we can use the native version
```suggestion
			if (!value.startsWith('"')) {
```
- the same goes for the other usages elsewhere

20k-ultra replied:
I've replaced any lodash implementations of startsWith to use standard lib startsWith

Page- commented on src/config/backends/raspberry-pi.ts:
```diff
@@ -0,0 +1,157 @@
+	}
+
+	public processConfigVarValue(key: string, value: string): string | string[] {
+		if (_.includes(RPiConfigBackend.arrayConfigKeys, key)) {
```
Same here, if we know it's an array we can use the native version
```suggestion
		if (RPiConfigBackend.arrayConfigKeys.includes(key)) {
```
- the same goes for the other usages elsewhere

20k-ultra replied:
I've replaced any lodash implementations of includes to use standard lib includes

Page- commented on src/config/backends/raspberry-pi.ts:
```diff
@@ -0,0 +1,157 @@
+						conf[key] = [];
+					}
+					const confArr = conf[key];
+					if (!_.isArray(confArr)) {
```
```suggestion
					if (!Array.isArray(confArr)) {
```

20k-ultra replied:
I've replaced all Lodash instance of Array checking with standard lib's version

Page- commented on src/config/backends/raspberry-pi.ts:
```diff
@@ -0,0 +1,157 @@
+	private static bootConfigPath = `${bootMountPoint}/config.txt`;
+
+	public static bootConfigVarRegex = new RegExp(
+		'(' + _.escapeRegExp(RPiConfigBackend.bootConfigVarPrefix) + ')(.+)',
```
Since you only seem to care about the second group you could make the first non-capturing like
```suggestion
		'(?:' + _.escapeRegExp(RPiConfigBackend.bootConfigVarPrefix) + ')(.+)',
```
and then use `$1` in the replacement instead

20k-ultra replied:
added

Page- commented on src/config/backends/extlinux.ts:
```diff
@@ -0,0 +1,262 @@
+	private fdtDirective = new FDTDirective();
+
+	public static bootConfigVarRegex = new RegExp(
+		'(' + _.escapeRegExp(ExtlinuxConfigBackend.bootConfigVarPrefix) + ')(.+)',
```
Since you only seem to care about the second group you could make the first non-capturing like
```suggestion
		'(?:' + _.escapeRegExp(ExtlinuxConfigBackend.bootConfigVarPrefix) + ')(.+)',
```
and then use `$1` in the replacement instead

20k-ultra replied:
added

Page- commented on src/config/backends/raspberry-pi.ts:
```diff
@@ -0,0 +1,157 @@
+import * as _ from 'lodash';
+import { fs } from 'mz';
```
This should be able to be
```suggestion
import { promises as fs } from 'fs';
```
now that the supervisor uses nodejs 12 and work the same, except using the in-built promise interface (it was added in nodejs 10)
- the same goes for other usage as well

20k-ultra replied:
I've created https://github.com/balena-io/balena-supervisor/issues/1384 to address this in another PR.

Page- commented on src/config/backends/extlinux.ts:
```diff
@@ -0,0 +1,262 @@
+	}
+
+	private static findDefaultLabel(file: ExtlinuxFile): string {
+		const defaultLabel = _.find(file.globals, (_v, l) => {
+			if (l === 'DEFAULT') {
+				return true;
+			}
+			return false;
+		});
```
I think this can be
```suggestion
		const defaultLabel = file.globals.DEFAULT;
```
?

20k-ultra replied:
Yes, and I was able to make it even simpler by just check if the value is falsey else return it. No need to find.

Page- commented on src/config/backends/extlinux.ts:
```diff
@@ -0,0 +1,262 @@
+		// Firstly split by line and filter any comments and empty lines
+		let lines = confStr.split(/\r?\n/);
+		lines = _.filter(lines, (l) => {
+			const trimmed = _.trimStart(l);
```
You could technically include this trim in the split above with
```ts
let lines = confStr.split(/\r?\n\s*/);
```
I'm not sure I'm a big fan of it, but I found it interesting to note.

And actually I just realized you could replace the entire filter with
```ts
let lines = confStr.split(/(?:\r?\n[\s#]*)+/);
```

20k-ultra replied:
Yup this works great!

Page- commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,86 @@
+	}
+
+	public value(opts: ConfigOptions): string {
+		if (opts.fdt.length === 0) {
+			throw new Error('Cannot set FDT of an empty value.');
+		}
+		return opts.fdt as string;
```
```suggestion
		if (typeof opts.fdt !== 'string') {
			throw new Error($errorMessage);
		}
		if (opts.fdt.length === 0) {
			throw new Error('Cannot set FDT of an empty value.');
		}
		return opts.fdt;
```
to avoid the need for the cast, especially as I think this can come from user input?

20k-ultra replied:
added

Page- commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,86 @@
+		});
+
+		// Apply new configurations values to the provided append line
+		_.each(opts, (value, key) => {
+			if (!value) {
+				// Example: rootwait (config without a value)
+				appendLine.push(`${key}`);
+			} else {
+				// Example: isolcpus=2,3 (config with a value)
+				appendLine.push(`${key}=${value}`);
+			}
+		});
+
+		// Return modified line as a string
+		return appendLine.join(' ');
```
I prefer a map/concat style as it makes it clear that we're mapping over/transforming the array in a direct 1-1 style
```suggestion
		return appendLine.concat(_.map(opts, (value, key) => {
			if (!value) {
				// Example: rootwait (config without a value)
				return `${key}`;
			} else {
				// Example: isolcpus=2,3 (config with a value)
				return `${key}=${value}`;
			}
		})).join(' ');
```

20k-ultra replied:
I agree and added this implementation.

Page- commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,86 @@
+
+	public value(opts: ConfigOptions, existingValue: string): string {
+		// Parse current append line and remove already set values so we don't have duplicates
+		const appendLine = _.filter(existingValue.split(' '), (entry) => {
```
```suggestion
		const appendLine = existingValue.split(' ').filter((entry) => {
```

Page- commented on src/config/backends/config-fs.ts:
```diff
@@ -0,0 +1,231 @@
+	}
+
+	private async loadConfiguredSsdt(config: ConfigfsConfig): Promise<void> {
+		if (_.isArray(config['ssdt'])) {
```
```suggestion
		if (Array.isArray(config['ssdt'])) {
```

Page- commented on src/config/backends/config-fs.ts:
```diff
@@ -0,0 +1,231 @@
+				`AML: ${oemId.trim()} ${oemTableId.trim()} (Rev ${oemRevision.trim()})`,
+			);
+		} catch (e) {
+			log.error(e);
```
Can you add a string message to go with `e`, ie
```suggestion
			log.error($errorMessage, e);
```
It makes it far easier to find where a log message is coming from in the case that `e` itself is not very useful, or could be coming from multiple places

20k-ultra commented:
I have added most of the changes commented except for the 2 that are currently unresolved. I also added 2 new tests files specifically for each directive. This let me find a lot of potential bugs so it was a great addition.

Page- commented on src/config/backends/raspberry-pi.ts:
```diff
@@ -0,0 +1,157 @@
+	}
+
+	public async setBootConfig(opts: ConfigOptions): Promise<void> {
+		let confStatements: string[] = [];
+
+		_.each(opts, (value, key) => {
+			if (key === 'initramfs') {
+				confStatements.push(`${key} ${value}`);
+			} else if (Array.isArray(value)) {
+				confStatements = confStatements.concat(
+					value.map((entry) => `${key}=${entry}`),
+				);
+			} else {
+				confStatements.push(`${key}=${value}`);
+			}
+		});
```
```suggestion
		const confStatements = _.flatMap(opts, (value, key) => {
			if (key === 'initramfs') {
				return `${key} ${value}`;
			} else if (Array.isArray(value)) {
				return value.map((entry) => `${key}=${entry}`);
			} else {
				return `${key}=${value}`;
			}
		});
```

20k-ultra replied:
added this change

Page- commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,166 @@
+						throw new AppendDirectiveError(
+							`One of the values being set contains an invalid character: [ value: ${value}, key: ${key} ]`,
+						);
+					} else if (!value) {
```
What are you looking for with `!value`? If it can be null then the `value.includes('=')` would need to be `value?.includes('=')`

20k-ultra replied:
good catch. I had thought includes would return false since that character is not included in null/undefined but actually includes is only for string prototype so it errors...

Page- commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,166 @@
+		// Parse current append line and remove whitelisted values
+		// We remove whitelisted values to avoid duplicates
+		const appendLine = existingValue.split(' ').filter((entry) => {
+			const lhs = entry.split('=');
```
It's worth limiting this to return just the first result since that's all we care about:
```suggestion
			const lhs = entry.split('=', 1);
```

Page- commented on src/config/backends/extlinux-file.ts:
```diff
@@ -0,0 +1,166 @@
+		return directives.APPEND.split(' ').reduce(
+			(configOpts: ConfigOptions, appendValue: string) => {
+				// Break this append config into key and value
+				const [KEY, VALUE = '', ...more] = appendValue.split('=');
+				if (!KEY) {
+					return configOpts; // No value to set so return
+				} else if (more.length > 0) {
```
You can limit to a max of 3 returned and also avoid building up the whole array of junk results by doing:
```suggestion
				const [KEY, VALUE = '', more] = appendValue.split('=', 3);
				if (!KEY) {
					return configOpts; // No value to set so return
				} else if (more != null) {
```
since `more` will either be a string or nullish:
```ts
'a=b='.split('=') // ["a", "b", ""]
'a=b'.split('=') // ["a", "b"]

'a=b==='.split('=') // ["a", "b", "", "", ""]
'a=b==='.split('=', 3) // ["a", "b", ""]

20k-ultra replied:
I like this philosophy of specifying the exact number of splits to return. I'll be sure to use it more. added.

Page- commented on src/config/backends/config-fs.ts:
```diff
@@ -0,0 +1,231 @@
+		if (Array.isArray(config['ssdt'])) {
+			log.info('Loading configured SSDTs');
+			for (const aml of config['ssdt']) {
+				await this.loadAML(aml);
```
Should this be done in serial or in parallel?

20k-ultra replied:
This is from existing functionality.

Page- commented on src/config/backends/config-fs.ts:
```diff
@@ -0,0 +1,231 @@
+			log.info(`Loading AML ${aml}`);
+			// we use `cat` here as this didn't work when using `cp` and all
+			// examples of this loading mechanism use `cat`.
+			await child_process.exec(
+				`cat ${amlSrcPath} > ${path.join(amlDstPath, 'aml')}`,
+			);
```
Any reason not to handle this in nodejs, eg
```ts
fs.createReadStream(amlSrcPath).on('error', ...).pipe(fs.createWriteStream(path.join(amlDstPath, 'aml'))).on('error', ...);
```
it just feel a bit weird to be calling out to another binary to handle this

CameronDiver replied:
@richbayliss should know the reason exactly, but I remember none of nodes tools seemed to work (streams, async and sync).
